### PR TITLE
Consent & local storage config flag

### PR DESCRIPTION
**Pull request by shandilya3**
## Description

 

-    **personalizationStorageEnabled** is set to true by default. Local storage is only created when the flag is enabled.
-  If a customer sets **personalizationStorageEnabled** to false at some point, local storage  is cleared.
-    consent can have three states:  **in**, **out**, **pending** 
-    when personalizationStorageEnabled is true and consent is **in**, in that case local storage is setup once the consent is resolved.
- in the case of defaultConsent set to **out**,  await consent promise is not resolved & the promise falls into catch block, where we clear the local storage if it existed before.
- in the case of defaultConsent set to **pending**, sendEvents calls are queued up and processed only when opted-in

Future work :
- [ ] will add additional test with  functional tests in the same PR 

 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [X] I have made any necessary test changes and all tests pass.
- [X] I have run the Sandbox successfully.

### CODE REVIEW

File: src/components/DecisioningEngine/index.js [hunk 1]
```diff
@@ -48,6 +48,26 @@ const createDecisioningEngine = ({ config, createNamespacedStorage }) => {
       },
       onComponentsRegistered(tools) {
         applyResponse = createApplyResponse(tools.lifecycle);
+        if (personalizationStorageEnabled) {
```
Comment (carterworks) [inline src/components/DecisioningEngine/index.js:51]: Why is the storaged created inside the `onComponentsRegistered()` callback?

File: src/components/DecisioningEngine/index.js [hunk 2]
```diff
@@ -48,6 +48,26 @@ const createDecisioningEngine = ({ config, createNamespacedStorage }) => {
...
+            .then(() => {
+              const eventRegistry = createEventRegistry({
+                storage: storage.persistent
+              });
+              eventContext.setCurrentEventRegistry(eventRegistry);
+            })
+            .catch(() => {
+              // If consent is rejected, we want to clear the local storage.
+              if (storage) {
+                storage.persistent.clear();
```
Comment (jasonwaters) [inline src/components/DecisioningEngine/index.js:66]: If a customer sets `personalizationStorageEnabled` to false at some point, local storage should also be cleared.  That way customer's won't disable it and wonder why they still see an entry in local storage.

File: src/components/DecisioningEngine/index.js [hunk 3]
```diff
@@ -48,6 +48,26 @@ const createDecisioningEngine = ({ config, createNamespacedStorage }) => {
       },
       onComponentsRegistered(tools) {
         applyResponse = createApplyResponse(tools.lifecycle);
+        if (personalizationStorageEnabled) {
+          const storage = createNamespacedStorage(
+            `${sanitizeOrgIdForCookieName(orgId)}.decisioning.`
+          );
+          consent
```
Comment (jasonwaters) [inline src/components/DecisioningEngine/index.js:55]: I think it's fine to have `.awaitConsent()` here in the `index.js` file.  But I wonder if it makes sense to move the listener within `eventRegistry` instead? 🤔 There are pros/cons.

**Review by jasonwaters — CHANGES_REQUESTED** [review]

File: src/components/DecisioningEngine/createEventContext.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  const toJSON = () => {
+    return currentEventRegistry.toJSON();
+  };
+
+  const clear = () => {
+    currentEventRegistry.clear();
+  };
+
+  const setCurrentEventRegistry = eventRegistry => {
+    currentEventRegistry = eventRegistry;
```
Comment (carterworks) [inline src/components/DecisioningEngine/createEventContext.js:38]: If a new event registry is set, should it copy over the events from the previous registry?

File: src/components/DecisioningEngine/createEventRegistry.js [hunk 5]
```diff
@@ -132,5 +139,16 @@ export default ({ storage }) => {
     return events[eventType][eventId];
   };
 
-  return { addExperienceEdgeEvent, addEvent, getEvent, toJSON: () => events };
+  const setStorage = newStorage => {
```
Comment (jasonwaters) [inline src/components/DecisioningEngine/createEventRegistry.js:142]: I think it's redundant to have both `updateStorage` and `setStorage` functions.

I recommend removing this version of `setStorage` and renaming the `updateStorage` method above to `setStorage`.  Implementation of `setStorage` can be modified to accept `newStorage`  for example...

```js
  let currentStorage;
  let restore;
  let save;
  let events;

  const setStorage = newStorage => {
    currentStorage = newStorage;

    restore = createRestoreStorage(currentStorage, STORAGE_KEY);
    save = createSaveStorage(
      currentStorage,
      STORAGE_KEY,
      createEventPruner(MAX_EVENT_RECORDS, RETENTION_PERIOD)
    );
    events = restore({});
  };

  setStorage(storage);
```

**Review by jasonwaters — COMMENTED** [review]
looks great @shandilya3 !  Just one suggestion.

**Review by jasonwaters — APPROVED** [review]

**Pull request merged**